README.md
# Mongo-dart - MongoDB driver for Dart programming language

[![Pub](https://img.shields.io/pub/v/mongo_dart.svg)](https://pub.dartlang.org/packages/mongo_dart)
[![Build Status](https://travis-ci.org/mongo-dart/mongo_dart.svg?branch=master)](https://travis-ci.org/mongo-dart/mongo_dart)

Server-side driver library for MongoDb implemented in pure Dart.

## Basic usage

### Obtaining connection

```dart

  var db = Db("mongodb://localhost:27017/mongo_dart-blog");
  await db.open();
```

### Querying

Method `find` returns stream of maps and accept query parameters, usually build by fluent API query builder
provided by [mongo_dart_query](https://github.com/vadimtsushko/mongo_dart_query) as top level getter `where`

```dart

  var coll = db.collection('user');
  await coll.find(where.lt("age", 18)).toList();
  
  //....
  
  await coll
      .find(where.gt("my_field", 995).sortBy('my_field'))
      .forEach((v) => print(v));

  //....
  
  await coll.find(where.sortBy('itemId').skip(300).limit(25)).toList();
  
```

Method `findOne` take the same parameter and returns `Future` of just one map (mongo document)

```dart

  val = await coll.findOne(where.eq("my_field", 17).fields(['str_field','my_field']));
```

Take notice in these samples that unlike mongo shell such parameters as projection (`fields`), `limit` and `skip`
are passed as part of regular query through query builder

### Inserting documents

```dart

  await usersCollection.insertAll([
    {'login': 'jdoe', 'name': 'John Doe', 'email': 'john@doe.com'},
    {'login': 'lsmith', 'name': 'Lucy Smith', 'email': 'lucy@smith.com'}
  ]);
```

### Updating documents

You can update whole document with method `save`

```dart

  var v1 = await coll.findOne({"name": "c"});
  v1["value"] = 31;
  await coll.save(v1);
```

or you can perform field level updates with method `update` and top level getter `modify` for ModifierBuilder fluent API

```dart

  coll.update(where.eq('name', 'Daniel Robinson'), modify.set('age', 31));

```

### Removing documents

```dart

  students.remove(where.id(id));
  /// or, to remove all documents from collection
  students.remove();

```

Simple app on base of [JSON ZIPS dataset](https://media.mongodb.org/zips.json)

```dart
import 'package:mongo_dart/mongo_dart.dart';

void main() async {
  void displayZip(Map zip) {
    print(
        'state: ${zip["state"]}, city: ${zip["city"]}, zip: ${zip["id"]}, population: ${zip["pop"]}');
  }
  var db =
      Db("mongodb://reader:vHm459fU@ds037468.mongolab.com:37468/samlple");
  var zips = db.collection('zip');
  await db.open();
  print('''
******************** Zips for state NY, with population between 14000 and 16000,
******************** reverse ordered by population''');
  await zips
      .find(where
          .eq('state', 'NY')
          .inRange('pop', 14000, 16000)
          .sortBy('pop', descending: true))
      .forEach(displayZip);
  print('\n******************** Find ZIP for code 78829 (BATESVILLE)');
  var batesville = await zips.findOne(where.eq('id', '78829'));
  displayZip(batesville);
  print('******************** Find 10 ZIP closest to BATESVILLE');
  await zips
      .find(where.near('loc', batesville["loc"]).limit(10))
      .forEach(displayZip);
  print('closing db');
  await db.close();
}
```

### Building aggregation queries

```dart
import 'package: mongo_dart/mongo_dart.dart';

main() async {
  final db = Db('mongodb://127.0.0.1/testdb');
  final pipeline = AggregationPipelineBuilder()
    .addStage(
      Match(where.eq('status', 'A').map['\$query']))
    .addStage(
      Group(
        id: Field('cust_id'),
        fields: {
          'total': Sum(Field('amount'))
        }
      )).build();
  final result =
    await DbCollection(db, 'orders')
      .aggregateToStream(pipeline).toList();
  result.forEach(print);
}
```

### Secure connection

You can connect using a secured tls/ssl connection in one of these two ways:

* setting the secure connection parameter to true in db.open()

```dart
    await db.open(secure: true);
```

* adding a query parameter => "tls=true" (or "ssl=true").

```dart
    var db = DB('mongodb://www.example.com:27017/test?tls=true&authSource=admin');
              or
    var db = DB('mongodb://www.example.com:27017/test?ssl=true&authSource=admin');
```

No certificates can be used.

### Atlas (MongoDb cloud service) connection

Atlas requires a tls connection, so now it is possible to connect to this cloud service.
When creating a cluster Atlas shows you three ways of connecting:
Mongo shell, driver and MongoDb Compass Application.
The connection string is in Seedlist Connection Format (starts with mongodb+srv://).
At present this driver does not support this connection string format.
You can do the following:
connect with the mongo shell to the address given by the site, for example:

```bash
mongo "mongodb+srv://cluster0.xtest.mongodb.net/<database name>" --username <your Atlas user>
```

The shell will ask you the password, enter it.

immediately after, the shell will show you the connection string in Standard Connection Format (starting with "mongodb://") in a line starting with "connecting to", for example.

```code
connecting to: mongodb://cluster0-shard-00-00.xtest.mongodb.net:27017,cluster0-shard-00-02.xtest.mongodb.net:27017,cluster0-shard-00-01.xtest.mongodb.net:27017/<database name>?authSource=admin&compressors=disabled&gssapiServiceName=mongodb&replicaSet=atlas-stcn2i-shard-0&ssl=true
```

Copy that string and add your user and password immediately after the "mongodb://" schema in the format "username:password@", for example

```code
mongodb://<your user>:<your password>@cluster0-shard-00-00.xtest.mongodb.net:27017,cluster0-shard-00-02.xtest.mongodb.net:27017,cluster0-shard-00-01.xtest.mongodb.net:27017/<database name>?authSource=admin&compressors=disabled&gssapiServiceName=mongodb&replicaSet=atlas-stcn2i-shard-0&ssl=true
```

Here we are, you can use the latter Connection String in the Db constructor

```dart
var db = Db("mongodb://dbUser:password@cluster0-shard-00-00.xtest.mongodb.net:27017,cluster0-shard-00-02.xtest.mongodb.net:27017,cluster0-shard-00-01.xtest.mongodb.net:27017/test-db?authSource=admin&compressors=disabled&gssapiServiceName=mongodb&replicaSet=atlas-stcn2i-shard-0&ssl=true");
```

### See also

* [API Doc](https://pub.dev/documentation/mongo_dart/latest/)

* [Feature check list](https://github.com/vadimtsushko/mongo_dart/blob/master/doc/feature_checklist.md)

* [Recent change notes](https://github.com/vadimtsushko/mongo_dart/blob/master/changelog.md)

* Additional [examples](https://github.com/vadimtsushko/mongo_dart/tree/master/example) and [tests](https://github.com/vadimtsushko/mongo_dart/tree/master/test)


CHANGELOG.md
# Changelog

## Recent change notes

### 0.4.2-dev.10.1

* Merged 0.4.1-dev.2.1 fix

### 0.4.2-dev.10

* Committment control - Experimental!!!

### 0.4.1

* Secure Connection
* The connection string now accepts more than one server.
  * Before: only mongodb://www.example.org/test.
  * Now it can be: mongodb://www.example.org,www1.example.org,www2.example.org/test
  * It is equivalent to: db.pool([mongodb://www.example.org/test, mongodb://www1.example.org/test, mongodb://www2.example.org/test]);  
* Added an "uriList" getter in "Db" class.

### 0.4.1-dev.2.2

* Lint clean-up

### 0.4.1-dev.2.1

* Fix - socket error on connect was not catched

### 0.4.1-dev.2

* Better management of socket error.

### 0.4.1-dev.1

Fixes problems with createIndex method.

Breaking changes:

* The name of the index is now created without the leading underscore (Ex. creating an index on key "a", in the old version created an index named "_a_1", now the name will be "a_1").
* The minimum dart sdk version required has been raised to 2.5.2

Changes List:

* New message (called ModernMessage). This has been tested for createIndex() only, and could be the base for moving all operations and commands from the old messages to this new one. Gradually ;-)
* A new method createIndex() has been created in the DbCollection class. This only works starting from 3.6. By default the result of the call is a little different form the old one (no getLastError() call is needed). For compatibility reasons I introduced a new parameter (modernReply), if set to false, it performs a getLastError() and returns the same way as the old method.
* The original createIndex method in Db class has been changed in order to call the new one (with modernReply: false) if the release is at least 3.6.
* The original MongoMessage has been split into two. The response part (Called MongoResponseMessage) now is the super class of MongoReplyMessage and MongoModernMessage.
* Some WriteConcern instances have been deprecated.
* A new "ServerStatus" message has been created. It is called when the connection is opened, and some values are stored if needed. The method is called ServerStatus() (in Db class) and returns a Map with all available status information. This method is only available from 3.6. on.
* An experimental "InsertOne" method has been created (only from 3.6).
* A MongoDbNamespace and a ReadPreference class have been created, mainly for future use.
* New tests.
* The Pedantic package has been included as a dev_dependency. The analysis.yaml has been updated to a more recent format, also including the pedantic package (default set of Google lint rules).

### 0.4.0-dev.1

* Upgraded to `mongo_dart_query` `0.4.1`. The update adds aggregation pipeline builder.

### 0.3.6

* Updates for Dart 2.5 compatibility. As of Dart 2.5, the `Socket` class
emits `Uint8List`, rather than plain `List<int>`, so some fixes were necessary
to allow the package to continue to function.
  * Courtesy of @a14n, thank you!
  * Resolves <https://github.com/mongo-dart/mongo_dart/issues/160>, among other issues.

### 0.3.5

* Upgraded `uuid` constraint to `^2.0.0`.

### 0.3.4

* Added `cursor` option to `DbCollection.aggregate`.
  * Courtesy of @jodinathan, thank you!
  * Resolves <https://github.com/mongo-dart/mongo_dart/issues/120>
  * See PR: <https://github.com/mongo-dart/mongo_dart/pull/139>
* Added "Tobe O" to authors in `pubspec.yaml`.

### 0.3.2

* Cast fixes, for compatibility with Dart 2.
* Moved from `dart:convert` usage to `package:dart2_constant`, for the sake
of backwards compatibility.

### 0.3.1

* Changed the signature of `DbCollection.remove`; its `selector` parameter is now required,
while `writeConcern` is now a named optional parameter.

```dart
Future remove(selector, {WriteConcern writeConcern});
```

Resolves [#73](https://github.com/mongo-dart/mongo_dart/issues/73).

### 0.3.0

* Strong mode compliance. Preparing for Dart 2.0

### 0.2.8

* Fix bad type in _Connection class

### 0.2.7

* Minor changes, commented out code removed
* Big chunk of tests was commented out in 0.2.6 by error. All tests restored.

### 0.2.6

* update for new sdk (with 1.17.0 [MongoMessageHandler] was broken in checked mode)
* upgrade to current version of `crypto` package, and it's new (Converter based) API

### 0.2.5+1

* Fixed markdown for pub.

### 0.2.5

* Moved mongo_dart project to a new organization on [GitHub](https://github.com/mongo-dart/mongo_dart).
* Authentication schemes now use Secure Random coming from `dart:math` since 1.14.  **Note that this version now required SDK version >= 1.14 because of this**
* Added sort by text score to the query builders.

### 0.2.5-beta

* SCRAM-SHA1 authentication scheme added by Joel Trottier-Hebert. Random string (salt) is generated

lib/src/database/db.dart
part of mongo_dart;

/// [WriteConcern] control the acknowledgment of write operations with various paramaters.
class WriteConcern {
  /// Denotes the Write Concern level that takes the following values ([int] or [String]):
  ///
  /// * -1 Disables all acknowledgment of write operations, and suppresses all errors, including network and socket errors.
  /// * 0: Disables basic acknowledgment of write operations, but returns information about socket exceptions and networking errors to the application.
  /// * 1: Provides acknowledgment of write operations on a standalone mongod or the primary in a replica set.
  /// * A number greater than 1: Guarantees that write operations have propagated successfully to the specified number of replica set members including the primary.
  /// * "majority": Confirms that write operations have propagated to the majority of configured replica set
  /// * A tag set: Fine-grained control over which replica set members must acknowledge a write operation
  final w;

  /// Specifies a timeout for this Write Concern in milliseconds, or infinite if equal to 0.
  final int wtimeout;

  /// Enables or disable fsync() operation before acknowledgement of the requested write operation.
  /// If [true], wait for mongod instance to write data to disk before returning.
  final bool fsync;

  /// Enables or disable journaling of the requested write operation before acknowledgement.
  /// If [true], wait for mongod instance to write data to the on-disk journal before returning.
  final bool j;

  /// Creates a WriteConcern object
  const WriteConcern({this.w, this.wtimeout, this.fsync, this.j});

  /// No exceptions are raised, even for network issues.
  @deprecated
  static const ERRORS_IGNORED =
      WriteConcern(w: -1, wtimeout: 0, fsync: false, j: false);

  /// Write operations that use this write concern will return as soon as the message is written to the socket.
  /// Exceptions are raised for network issues, but not server errors.
  static const UNACKNOWLEDGED =
      WriteConcern(w: 0, wtimeout: 0, fsync: false, j: false);

  /// Write operations that use this write concern will wait for acknowledgement from the primary server before returning.
  /// Exceptions are raised for network issues, and server errors.
  static const ACKNOWLEDGED =
      WriteConcern(w: 1, wtimeout: 0, fsync: false, j: false);

  /// Exceptions are raised for network issues, and server errors; waits for at least 2 servers for the write operation.
  static const REPLICA_ACKNOWLEDGED =
      WriteConcern(w: 2, wtimeout: 0, fsync: false, j: false);

  /// Exceptions are raised for network issues, and server errors; the write operation waits for the server to flush
  /// the data to disk.
  @deprecated
  static const FSYNCED = WriteConcern(w: 1, wtimeout: 0, fsync: true, j: false);

  /// Exceptions are raised for network issues, and server errors; the write operation waits for the server to
  /// group commit to the journal file on disk.
  static const JOURNALED =
      WriteConcern(w: 1, wtimeout: 0, fsync: false, j: true);

  /// Exceptions are raised for network issues, and server errors; waits on a majority of servers for the write operation.
  static const MAJORITY =
      WriteConcern(w: 'majority', wtimeout: 0, fsync: false, j: false);

  /// Gets the getlasterror command for this write concern.
  Map<String, dynamic> get command {
    var map = <String, dynamic>{};
    map['getlasterror'] = 1;
    if (w != null) {
      map['w'] = w;
    }
    if (wtimeout != null) {
      map['wtimeout'] = wtimeout;
    }
    if (fsync != null) {
      map['fsync'] = fsync;
    }
    if (j != null) {
      map['j'] = j;
    }
    return map;
  }

  /// To be used starting with journaled engines (Only Wired Tiger, Journal Only)
  /// For inMemoryEngine the J option is ignored
  ///
  /// We can use before 4.2 testing if the journal is active
  /// (in this case fsync doesn't make any sense, taken from mongodb Jira:
  /// "fsync means sync using a journal if present otherwise the datafiles")
  /// In 4.0 journal cannot be disabled on wiredTiger engine
  /// In 4.2 only wiredTiger can be used
  Map<String, Object> asMap(ServerStatus serverStatus) {
    var ret = <String, Object>{};
    if (w != null) {
      ret['w'] = w;
    }
    if (wtimeout != null) {
      ret['wtimeout'] = wtimeout;
    }
    if (serverStatus.isPersistent) {
      if (j != null) {
        ret['j'] = j;
      }
      if (fsync != null && j != true) {
        if (serverStatus.isJournaled) {
          ret['j'] = fsync;
        } else {
          ret['fsync'] = fsync;
        }
      }
    }
    return ret;
  }
}

class _UriParameters {
  static const authMechanism = 'authMechanism';
  static const authSource = 'authSource';
  static const tls = 'tls';
  static const ssl = 'ssl';
}

class Db {
  static const mongoDefaultPort = 27017;
  final _log = Logger('Db');
  final List<String> _uriList = <String>[];

  State state = State.INIT;
  String databaseName;
  String _debugInfo;
  Db authSourceDb;
  _ConnectionManager _connectionManager;

  _Connection get _masterConnection => _connectionManager.masterConnection;

  _Connection get _masterConnectionVerified =>
      _connectionManager.masterConnectionVerified;
  WriteConcern _writeConcern;
  AuthenticationScheme _authenticationScheme;
  ReadPreference readPreference = ReadPreference.primary;

  @override
  String toString() => 'Db($databaseName,$_debugInfo)';

  /// Db constructor expects [valid mongodb URI](https://docs.mongodb.com/manual/reference/connection-string/).
  /// For example next code points to local mongodb server on default mongodb port, database *testdb*
  ///```dart
  ///     var db = new Db('mongodb://127.0.0.1/testdb');
  ///```
  /// And that code direct to MongoLab server on 37637 port, database *testdb*, username *dart*, password *test*
  ///```dart
  ///     var db = new Db('mongodb://dart:test@ds037637-a.mongolab.com:37637/objectory_blog');
  ///```
  Db(String uriString, [this._debugInfo]) {
    if (uriString.contains(',')) {
      _uriList.addAll(splitHosts(uriString));
    } else {
      _uriList.add(uriString);
    }
  }

  Db.pool(List<String> uriList, [this._debugInfo]) {
    _uriList.addAll(uriList);
  }

  Db._authDb(this.databaseName);

  /// This method allow to create a Db object both with the Standard
  /// Connection String Format (`mongodb://`) or with the DNS Seedlist
  /// Connection Format (`mongodb+srv://`).
  /// The former has the format:
  /// mongodb://[username:password@]host1[:port1]
  ///      [,...hostN[:portN]][/[defaultauthdb][?options]]
  /// The latter is available from version 3.6. The format is:
  /// mongodb+srv://[username:password@]host1[:port1]
  ///      [/[databaseName][?options]]
  /// More info are available [here](https://docs.mongodb.com/manual/reference/connection-string/)
  ///
  /// This is an asynchronous constructor.
  /// In order to resolve the Seedlist, a call to a DNS server is needed
  /// If the DNS server is unreachable, the constructor throws an error.
  static Future<Db> create(String uriString, [String _debugInfo]) async {
    if (uriString.startsWith('mongodb://')) {
      return Db(uriString, _debugInfo);
    } else if (uriString.startsWith('mongodb+srv://')) {
      var uriList = await decodeDnsSeedlist(Uri.parse(uriString));
      return Db.pool(uriList, _debugInfo);
    } else {
      throw MongoDartError(
          'The only valid schemas for Db are: "mongodb" and "mongodb+srv".');
    }
  }

  WriteConcern get writeConcern => _writeConcern;

  _Connection get masterConnection => _connectionManager.masterConnection;

  List<String> get uriList => _uriList.toList();

  ServerConfig _parseUri(String uriString, {bool isSecure}) {
    isSecure ??= false;
    var uri = Uri.parse(uriString);

    if (uri.scheme != 'mongodb') {
      throw MongoDartError('Invalid scheme in uri: $uriString ${uri.scheme}');
    }

    uri.queryParameters.forEach((String queryParam, String value) {
      if (queryParam == _UriParameters.authMechanism) {
        selectAuthenticationMechanism(value);
      }

      if (queryParam == _UriParameters.authSource) {
        authSourceDb = Db._authDb(value);
      }

      if ((queryParam == _UriParameters.tls ||
              queryParam == _UriParameters.ssl) &&
          value == 'true') {
        isSecure = true;
      }
    });

    var serverConfig = ServerConfig(
        uri.host ?? '127.0.0.1', uri.port ?? mongoDefaultPort, isSecure);

    if (serverConfig.port == 0) {
      serverConfig.port = mongoDefaultPort;
    }

    if (uri.userInfo.isNotEmpty) {
      var userInfo = uri.userInfo.split(':');

      if (userInfo.length != 2) {
        throw MongoDartError('Invalid format of userInfo field: $uri.userInfo');
      }

      serverConfig.userName = Uri.decodeComponent(userInfo[0]);
      serverConfig.password = Uri.decodeComponent(userInfo[1]);
    }

    if (uri.path.isNotEmpty) {
      databaseName = uri.path.replaceAll('/', '');
    }

    return serverConfig;
  }

  void selectAuthenticationMechanism(String authenticationSchemeName) {
    if (authenticationSchemeName == ScramSha1Authenticator.name) {
      _authenticationScheme = AuthenticationScheme.SCRAM_SHA_1;
    } else if (authenticationSchemeName == MongoDbCRAuthenticator.name) {
      _authenticationScheme = AuthenticationScheme.MONGODB_CR;
    } else {
      throw MongoDartError(
          'Provided authentication scheme is not supported : $authenticationSchemeName');
    }
  }

  DbCollection collection(String collectionName) {
    return DbCollection(this, collectionName);
  }

  Future<MongoReplyMessage> queryMessage(MongoMessage queryMessage,
      {_Connection connection}) {
    return Future.sync(() {
      if (state != State.OPEN) {
        throw MongoDartError('Db is in the wrong state: $state');
      }

      connection ??= _masterConnectionVerified;

      return connection.query(queryMessage);
    });
  }

  void executeMessage(MongoMessage message, WriteConcern writeConcern,
      {_Connection connection}) {
    if (state != State.OPEN) {
      throw MongoDartError('DB is not open. $state');
    }

    connection ??= _masterConnectionVerified;

    writeConcern ??= _writeConcern;

    connection.execute(message, writeConcern == WriteConcern.ERRORS_IGNORED);
  }

  Future<Map<String, Object>> executeModernMessage(MongoModernMessage message,
      {_Connection connection}) async {
    if (state != State.OPEN) {
      throw MongoDartError('DB is not open. $state');
    }

    connection ??= _masterConnectionVerified;

    var response = await connection.executeModernMessage(message);

    var section = response.sections.firstWhere((Section _section) =>
        _section.payloadType == MongoModernMessage.basePayloadType);
    return section.payload.content;
  }

  Future open(
      {WriteConcern writeConcern = WriteConcern.ACKNOWLEDGED,
      bool secure = false}) async {
    if (state == State.OPENING) {
      throw MongoDartError('Attempt to open db in state $state');
    }

    state = State.OPENING;
    _writeConcern = writeConcern;
    _connectionManager = _ConnectionManager(this);

    for (var uri in _uriList) {
      _connectionManager.addConnection(_parseUri(uri, isSecure: secure));
    }
  
    return _connectionManager.open(writeConcern);
  }

  Future<Map<String, dynamic>> executeDbCommand(MongoMessage message,
      {_Connection connection}) async {
    connection ??= _masterConnectionVerified;

    var result = Completer<Map<String, dynamic>>();

    var replyMessage = await connection.query(message);
    var firstRepliedDocument = replyMessage.documents[0];
    var errorMessage = '';

    if (replyMessage.documents.isEmpty) {
      errorMessage =
          'Error executing Db command, documents are empty $replyMessage';

      print('Error: $errorMessage');

      var m = <String, dynamic>{};
      m['errmsg'] = errorMessage;

      result.completeError(m);
    } else if (documentIsNotAnError(firstRepliedDocument)) {
      result.complete(firstRepliedDocument);
    } else {
      result.completeError(firstRepliedDocument);
    }
    return result.future;
  }

  bool documentIsNotAnError(firstRepliedDocument) =>
      firstRepliedDocument['ok'] == 1.0 && firstRepliedDocument['err'] == null;

  Future<bool> dropCollection(String collectionName) async {
    var collectionInfos = await getCollectionInfos({'name': collectionName});

    if (collectionInfos.length == 1) {
      return executeDbCommand(
              DbCommand.createDropCollectionCommand(this, collectionName))
          .then((_) => true);
    }

    return true;
  }

  ///   Drop current database
  Future drop() {
    return executeDbCommand(DbCommand.createDropDatabaseCommand(this));
  }

  Future<Map<String, dynamic>> removeFromCollection(String collectionName,
      [Map<String, dynamic> selector = const {}, WriteConcern writeConcern]) {
    return Future.sync(() {
      executeMessage(
          MongoRemoveMessage('$databaseName.$collectionName', selector),
          writeConcern);
      return _getAcknowledgement(writeConcern: writeConcern);
    });
  }

  Future<Map<String, dynamic>> getLastError([WriteConcern writeConcern]) {
    writeConcern ??= _writeConcern;
    return executeDbCommand(
        DbCommand.createGetLastErrorCommand(this, writeConcern));
  }

  Future<Map<String, dynamic>> getNonce({_Connection connection}) {
    return executeDbCommand(DbCommand.createGetNonceCommand(this),
        connection: connection);
  }

  Future<Map<String, dynamic>> getBuildInfo({_Connection connection}) {
    return executeDbCommand(DbCommand.createBuildInfoCommand(this),
        connection: connection);
  }

  Future<Map<String, dynamic>> isMaster({_Connection connection}) {
    return executeDbCommand(DbCommand.createIsMasterCommand(this),
        connection: connection);
  }

  Future<Map<String, dynamic>> wait() {
    return getLastError();
  }

  Future close() {
    _log.fine(() => '$this closed');
    state = State.CLOSED;
    var _cm = _connectionManager;
    _connectionManager = null;
    return _cm.close();
  }

  /// Analogue to shell's `show dbs`. Helper for `listDatabases` mongodb command.
  Future<List> listDatabases() async {
    var commandResult = await executeDbCommand(
        DbCommand.createQueryAdminCommand({'listDatabases': 1}));

    var result = [];

    for (var each in commandResult['databases']) {
      result.add(each['name']);
    }

    return result;
  }

  Stream<Map<String, dynamic>> _listCollectionsCursor(
      [Map<String, dynamic> filter = const {}]) {
    if (_masterConnection.serverCapabilities.listCollections) {
      return ListCollectionsCursor(this, filter).stream;
    } else {
      // Using system collections (pre v3.0 API)
      var selector = <String, dynamic>{};
      // If we are limiting the access to a specific collection name
      if (filter.containsKey('name')) {
        selector['name'] = "${databaseName}.${filter['name']}";
      }
      return Cursor(
              this,
              DbCollection(this, DbCommand.SYSTEM_NAMESPACE_COLLECTION),
              selector)
          .stream;
    }
  }

  /// This method uses system collections and therefore do not work on MongoDB v3.0 with and upward
  /// with WiredTiger
  /// Use `getCollectionInfos` instead
  @deprecated
  Stream<Map<String, dynamic>> collectionsInfoCursor([String collectionName]) {
    return _collectionsInfoCursor(collectionName);
  }

  Stream<Map<String, dynamic>> _collectionsInfoCursor([String collectionName]) {
    var selector = <String, dynamic>{};
    // If we are limiting the access to a specific collection name
    if (collectionName != null) {
      selector['name'] = '${databaseName}.$collectionName';
    }
    // Return Cursor
    return Cursor(this,
            DbCollection(this, DbCommand.SYSTEM_NAMESPACE_COLLECTION), selector)
        .stream;
  }

  /// Analogue to shell's `show collections`
  /// This method uses system collections and therefore do not work on MongoDB v3.0 with and upward
  /// with WiredTiger
  /// Use `getCollectionNames` instead
  @deprecated
  Future<List<String>> listCollections() {
    return _collectionsInfoCursor()
        .map((map) => map['name']?.toString()?.split('.'))
        .where((arr) => arr.length == 2)
        .map((arr) => arr.last)
        .toList();
  }

  Future<List<Map<String, dynamic>>> getCollectionInfos(
      [Map<String, dynamic> filter = const {}]) {
    return _listCollectionsCursor(filter).toList();
  }

  Future<List<String>> getCollectionNames(
      [Map<String, dynamic> filter = const {}]) {
    return _listCollectionsCursor(filter)
        .map((map) => map['name']?.toString())
        .toList();
  }

  Future<bool> authenticate(String userName, String password,
      {_Connection connection}) async {
    var credential = UsernamePasswordCredential()
      ..username = userName
      ..password = password;

    var authenticator =
        createAuthenticator(_authenticationScheme, this, credential);

    await authenticator.authenticate(connection ?? _masterConnection);

    return true;
  }

  /// This method uses system collections and therefore do not work on MongoDB v3.0 with and upward
  /// with WiredTiger
  /// Use `DbCollection.getIndexes()` instead
  @deprecated
  Future<List> indexInformation([String collectionName]) {
    var selector = {};

    if (collectionName != null) {
      selector['ns'] = '$databaseName.$collectionName';
    }

    return Cursor(this, DbCollection(this, DbCommand.SYSTEM_INDEX_COLLECTION),
            selector)
        .stream
        .toList();
  }

  String _createIndexName(Map<String, dynamic> keys) {
    var name = '';

    keys.forEach((key, value) {
      if (name.isEmpty) {
        name = '${key}_$value';
      } else {
        name = '${name}_${key}_$value';
      }
    });

    return name;
  }

  Future<Map<String, dynamic>> createIndex(String collectionName,
      {String key,
      Map<String, dynamic> keys,
      bool unique,
      bool sparse,
      bool background,
      bool dropDups,
      Map<String, dynamic> partialFilterExpression,
      String name}) {
    if (_masterConnection.serverCapabilities.supportsOpMsg) {
      return collection(collectionName).createIndex(
          key: key,
          keys: keys,
          unique: unique,
          sparse: sparse,
          background: background,
          dropDups: dropDups,
          partialFilterExpression: partialFilterExpression,
          name: name,
          modernReply: false);
    }
    return Future.sync(() async {
      var selector = <String, dynamic>{};
      selector['ns'] = '$databaseName.$collectionName';
      keys = _setKeys(key, keys);
      selector['key'] = keys;

      if (unique == true) {
        selector['unique'] = true;
      } else {
        selector['unique'] = false;
      }
      if (sparse == true) {
        selector['sparse'] = true;
      }
      if (background == true) {
        selector['background'] = true;
      }
      if (dropDups == true) {
        selector['dropDups'] = true;
      }
      if (partialFilterExpression != null) {
        selector['partialFilterExpression'] = partialFilterExpression;
      }
      name ??= _createIndexName(keys);
      selector['name'] = name;
      var insertMessage = MongoInsertMessage(
          '$databaseName.${DbCommand.SYSTEM_INDEX_COLLECTION}', [selector]);
      await executeMessage(insertMessage, _writeConcern);
      return getLastError();
    });
  }

  Map<String, dynamic> _setKeys(String key, Map<String, dynamic> keys) {
    if (key != null && keys != null) {
      throw ArgumentError('Only one parameter must be set: key or keys');
    }

    if (key != null) {
      keys = {};
      keys['$key'] = 1;
    }

    if (keys == null) {
      throw ArgumentError('key or keys parameter must be set');
    }

    return keys;
  }

  Future ensureIndex(String collectionName,
      {String key,
      Map<String, dynamic> keys,
      bool unique,
      bool sparse,
      bool background,
      bool dropDups,
      Map<String, dynamic> partialFilterExpression,
      String name}) async {
    keys = _setKeys(key, keys);
    var indexInfos = await collection(collectionName).getIndexes();

    name ??= _createIndexName(keys);

    if (indexInfos.any((info) => info['name'] == name) ||
        // For compatibility reasons, old indexes where created with
        // a leading underscore
        indexInfos.any((info) => info['name'] == '_$name')) {
      return {'ok': 1.0, 'result': 'index preexists'};
    }

    var createdIndex = await createIndex(collectionName,
        keys: keys,
        unique: unique,
        sparse: sparse,
        background: background,
        dropDups: dropDups,
        partialFilterExpression: partialFilterExpression,
        name: name);

    return createdIndex;
  }

  /// This method return the status information on the
  /// connection.
  ///
  /// Only works from version 3.6
  Future<Map<String, Object>> serverStatus(
      {Map<String, Object> options}) async {
    if (!_masterConnection.serverCapabilities.supportsOpMsg) {
      return <String, Object>{};
    }
    var operation = ServerStatusOperation(this, options: options);
    return operation.execute();
  }

  Future<Map<String, dynamic>> _getAcknowledgement(
      {WriteConcern writeConcern}) {
    writeConcern ??= _writeConcern;

    if (writeConcern == WriteConcern.ERRORS_IGNORED) {
      return Future.value({'ok': 1.0});
    } else {
      return getLastError(writeConcern);
    }
  }
}
